x-pack/solutions/observability/plugins/observability_agent_builder/server/routes/ai_insights/apm_error/fetch_distributed_trace.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { IScopedClusterClient, Logger } from '@kbn/core/server';
+import type { APMIndices } from '@kbn/apm-sources-access-plugin/server';
+import { accessKnownApmEventFields } from '@kbn/apm-data-access-plugin/server/utils';
+import type { UnflattenedApmEvent } from '@kbn/apm-data-access-plugin/server/utils/utility_types';
+import { getTypedSearch } from '../../../utils/get_typed_search';
+import { termFilter, timeRangeFilter } from '../../../utils/dsl_filters';
+
+interface ServiceAggregate {
+  'service.name': string;
+  count: number;
+  error_count: number;
+}
+
+export interface DistributedTrace {
+  traceDocuments: Array<UnflattenedApmEvent>;
+  traceServiceAggregates: ServiceAggregate[];
+}
+
+export async function fetchDistributedTrace({
+  esClient,
+  apmIndices,
+  traceId,
+  start,
+  end,
+  logger,
+}: {
+  esClient: IScopedClusterClient;
+  apmIndices: APMIndices;
+  traceId: string;
+  start: number;
+  end: number;
+  logger: Logger;
+}): Promise<DistributedTrace> {
+  const search = getTypedSearch(esClient.asCurrentUser);
+  const indices = [apmIndices.transaction, apmIndices.span, apmIndices.error].join(',');
+
+  const traceResponse = await search({
+    index: indices,
+    size: 100,
+    track_total_hits: false,
```

Comment:
nit: if we want to know if it is a partial trace (aka if there are more than 100 entries), we need to count to `size + 1`.

```suggestion
    track_total_hits: 101,
```

then we can determine if the trace is partial:

```ts 
const isPartialTrace = response.hits.total.relation === 'gte';
```